### PR TITLE
ASAN fixes

### PR DESCRIPTION
**Pull request by maleadt**

### CODE REVIEW

File: src/julia_internal.h [hunk 1]
```diff
@@ -291,7 +291,7 @@ jl_value_t *jl_permbox32(jl_datatype_t *t, int32_t x);
 jl_value_t *jl_permbox64(jl_datatype_t *t, int64_t x);
 jl_svec_t *jl_perm_symsvec(size_t n, ...);
 
-#if !defined(__clang_analyzer__) // this sizeof(__VA_ARGS__) trick can't be computed until C11, but only the analyzer seems to care
+#if !defined(_COMPILER_CLANG_) // this sizeof(__VA_ARGS__) trick can't be computed until C11, but only Clang seems to care (when analyzing or using ASAN)
```
Comment (Keno) [inline src/julia_internal.h]: I would prefer to keep this protection if at all possible. Not sure why ASAN breaks it but if it works fine with regular clang, can we just disable it under asan? Alternatively, we could set our default C standard to C11.

**Comment by Keno** [comment]
I'm confused why the datalayout commit is necessary. All of these have nfields set to zero, so it should never attempt to load any pointer past the end of the declared struct. Which code is doing this reading?

**Comment by maleadt** [comment]
> nfields set to zero

The number of pointers isn't set to 0, so this adds one:
https://github.com/JuliaLang/julia/blob/08ebedcbb86b140baa15f31097981e7800f15d0e/src/staticdata.c#L873-L874

**Comment by Keno** [comment]
I think the idea was that if first_ptr is `-1`, then those trailing bytes aren't allocated.

File: src/datatype.c [hunk 2]
```diff
@@ -360,19 +360,26 @@ void jl_compute_field_offsets(jl_datatype_t *st)
         // if we have no fields, we can trivially skip the rest
         if (st == jl_symbol_type || st == jl_string_type) {
             // opaque layout - heap-allocated blob
-            static const jl_datatype_layout_t opaque_byte_layout = {0, 1, -1, 1, 0, 0};
```
Comment (vtjnash) [inline src/datatype.c]:
```suggestion
static const struct {jl_datatype_layout_t fields; int8_t pointers} opaque_byte_layout = {{0, 1, -1, 1, 0, 0}, -1};
st->layout = &opaque_byte_layout.fields;
```

But seems like the reading this field would imply a problem is going to happen somewhere else shortly thereafter?

**Comment by maleadt** [comment]
> I think the idea was that if first_ptr is `-1`, then those trailing bytes aren't allocated.

Ah, OK. Yeah, that wasn't clear to me at all. I'll have a look at avoiding those reads then.

**Comment by maleadt** [comment]
I don't really get the layout code (i.e. the specific semantics of nfields/npointers/first_ptr, @vtjnash?), so I just implemented https://github.com/JuliaLang/julia/pull/34389#issuecomment-574787963.

**Review by Keno — APPROVED** [review]

**Pull request merged**